components/camel-bindy/src/main/java/org/apache/camel/dataformat/bindy/kvp/BindyKeyValuePairDataFormat.java

```diff
@@ -88,55 +92,73 @@ public void marshal(Exchange exchange, Object body, OutputStream outputStream) t
         InputStreamReader in = new InputStreamReader(inputStream, IOHelper.getCharsetName(exchange));
 
-        // Scanner is used to read big file
-        Scanner scanner = new Scanner(in);
+        // Use a Stream to stream a file across
+        try (Stream<String> lines = new BufferedReader(in).lines()) {
+            // Retrieve the pair separator defined to split the record
+            ObjectHelper.notNull(factory.getPairSeparator(), "The pair separator property of the annotation @Message");
+            String separator = factory.getPairSeparator();
+            AtomicInteger count = new AtomicInteger(0);
+
+            try {
+                lines.forEachOrdered(line -> {
+                    consumeFile(factory, models, lists, separator, count, line);
+                });
+            } catch (WrappedException e) {
+                throw e.getWrappedException();
+            }
+
+            // BigIntegerFormatFactory if models list is empty or not
+            // If this is the case (correspond to an empty stream, ...)
+            if (models.size() == 0) {
+                throw new java.lang.IllegalArgumentException("No records have been defined in the CSV");
+            } else {
+                return extractUnmarshalResult(models);
+            }
 
-        // Retrieve the pair separator defined to split the record
-        ObjectHelper.notNull(factory.getPairSeparator(), "The pair separator property of the annotation @Message");
-        String separator = factory.getPairSeparator();
+        } finally {
+            IOHelper.close(in, "in", LOG);
+        }
+    }
 
-        int count = 0;
+    private void consumeFile(BindyKeyValuePairFactory factory, List<Map<String, Object>> models, Map<String, List<Object>> lists, String separator, AtomicInteger count, String line) {
         try {
-            while (scanner.hasNextLine()) {
-                // Read the line
-                String line = scanner.nextLine().trim();
-
-                if (ObjectHelper.isEmpty(line)) {
-                    // skip if line is empty
-                    continue;
-                }
+            // Trim the line coming in to remove any trailing whitespace
+            String trimmedLine = line.trim();
 
+            if (!ObjectHelper.isEmpty(trimmedLine)) {
                 // Increment counter
-                count++;
+                count.incrementAndGet();
+                // Pojos of the model
+                Map<String, Object> model;
 
                 // Create POJO
                 model = factory.factory();
 
                 // Split the message according to the pair separator defined in
                 // annotated class @Message
-                List<String> result = Arrays.asList(line.split(separator));
+                // Explicitly replace any occurrence of the Unicode new line character.
+                List<String> result = Arrays.stream(line.split(separator))
+                        .map(x -> x.replace("\u0085", ""))
```

Comment:
could you explain this a little bit more?

Reply:
I'll add more details in comments, but the first part of simply reading the line doesn't get us away from the fact that the NEL character can still be present as a part of the key-value pair.  Explicitly replacing it here removes that concern.